src/cargo/core/compiler/mod.rs

```diff
@@ -144,17 +146,16 @@ fn compile<'cfg>(
             };
             work.then(link_targets(cx, unit, false)?)
         } else {
-            let work = if unit.show_warnings(bcx.config) {
-                replay_output_cache(
-                    unit.pkg.package_id(),
-                    &unit.target,
-                    cx.files().message_cache_path(unit),
-                    cx.bcx.build_config.message_format,
-                    cx.bcx.config.shell().err_supports_color(),
-                )
-            } else {
-                Work::noop()
-            };
+            // We always replay the output cache,
+            // since it might contain future-incompat-report messages
+            let work = replay_output_cache(
```

Comment:
This runs the risk of slowing down incremental compiles since this may do nontrivial work for all upstream crates (which can be a lot), can you try to do some testing locally to see what the impact is?

Reply:
I tested with `actix-web` (211 dependencies) - without running any kind of precise benchmarking, it looks like this may have increased the runtime of `cargo build` (with a full cache) by a few hundredths of a second (0.08 to 0.09 seconds before, 0.11 to 0.12 seconds after).

I think this worst case for this change would be:

1. Running `cargo build -vv` on a compilation graph with a significant number of normal upstream warnings.
2. Repeatedly running `cargo build` with the cache prepared by `cargo build -vv`, which cauess Cargo to process and discard a large number of cached diagnostic messages.

However, this seems like a pretty unlikely scenarion, and I haven't been able to measure any kind of significant impact locally (testing with `actix-web` again).

Reply:
Drive by comment, what system was this testing on? Can we check on windows?

Reply:
I tested on Arch Linux

Reply:
I agree that if we stick with this scheme that we'll want to test on Windows which typically has a slower filesystem than other platforms. Additionally the case I'm worried about is not when upstream compiles have output but rather they have no output so Cargo opens and closes a whole lot of files rapidly which can increase the latency of Cargo itself.

Reply:
Yes - we still want to display a note when running `cargo build` without any arguments.

Reply:
This is another reason I would prefer to remove the flag on `cargo build` in the initial implementation. I don't believe this change is necessary if it's a separate subcommand, right? If so that would make the performance issue moot.

Reply:
Hm ok I believe I'm wrong that this would change if there were only a subcommand. We'd still need to scrape files from past cached compilations to have the same report every time.

I don't think any testing of this has happened on Windows yet though?